volumes/short_term_counting_program/move_match/speedvol_stats.sql

```diff
@@ -0,0 +1,251 @@
+    FROM spd_vol
+    GROUP BY 
+        spd_vol.arterycode,
+        spd_vol.count_info_id,
+        spd_vol.direction,
+        spd_vol.location,
+        spd_vol.datetime_bin::date,
+        spd_vol.speed_kph
+    WINDOW
+        cumulative AS (PARTITION BY spd_vol.count_info_id ORDER BY spd_vol.speed_kph RANGE UNBOUNDED PRECEDING),
+        _all AS (PARTITION BY spd_vol.count_info_id)
+),
+
+/*
+SUM(SUM()) Explained
+There are two lines in the CTE above that use a nested aggregate function within a WINDOW function.
+A WINDOW function allows calculations to be performed on subsets of data (or WINDOWs of data) without collapsing the data into aggregated forms (like GROUP BY).
+In this CTE, two WINDOWs are used: 
+    cumulative: creates windows based on count_info_id and speed bin
+    _all: creates windows based on count_info_id
+Here is an explanation of the first SUM(SUM()) line:
+    Part 1 - SUM(SUM(spd_vol.volume_15min)) OVER cumulative: 
+        The inside SUM() adds up 15 minute volume data (resulting in daily volumes by speed bin).
+        The outside SUM() adds up the bins over the cumulative window.
+        In plain language: sum daily totals by speed bin, then calculate a cumulative sum for ascending speed bins.
+    Part 2 - / (SUM(SUM(spd_vol.volume_15min)) OVER _all + 0.00000001):
+        The inside SUM() adds up 15 minute volume data (resulting daily volumes by speed bin).
+        The outside SUM() adds up all of the speed bin volumes (resulting in daily volumes).
+        In plain language: calculate daily volumes.
+    Since Part 1 is divided by Part 2, you end up with cumulative proportions of daily volume by speed bin.
+Here is an explanation of the second SUM(SUM()) line:
+    SUM(spd_vol.volume_15min) / (SUM(SUM(spd_vol.volume_15min)) OVER _all + 0.00000001) AS pct:
+    Calculate daily volumes by speed bin and divide it by...
+    Daily total volumes (15 minute volume data are aggregated to speed bins in the inside sum, then they are totalled in the outside sum).
+    You end up with the proportion of daily volume in each speed bin.
+*/
+
+-- Calculate 15th percentile speed
+per_15 AS (
+    SELECT DISTINCT ON (spd_bin_vol.count_info_id)
+        spd_bin_vol.count_info_id,
+        spd_bin_vol.arterycode,
+        spd_bin_vol.dt,
+        spd_bin_vol.cum_pct,
+        spd_bin_vol.pct,
+        (
+            LOWER(spd_bin_vol.speed_kph) -- lower speed bound of bin
+            + (UPPER(spd_bin_vol.speed_kph) - LOWER(spd_bin_vol.speed_kph))
+            * (
+                0.15 -- multiplied by the desired percentile (for which you want speed)
+                - ( -- minus
+                    spd_bin_vol.cum_pct -- the cumulative % for that bin
+                    - spd_bin_vol.pct -- munis the non-cumulative % from that bin
+                )
+                ) / (spd_bin_vol.pct + 0.00000001 -- all dividded by the bin volume % plus that decimal to avoid a divide by zero error
```

Comment:
Filtering out zeros or using COALESCE would be better than adding to the denominator, which impacts the result ever so slightly.

Reply:
Accomplished by excluding bins with pct = 0. These could not include the desired percentile anyway because they have the same upper bound as the previous bin which would be chosen instead.